### ISSUE TITLE
Roadmap 1.0.0

### ISSUE THREAD

**Issue by Yomguithereal** [Yomguithereal/baobab#139]
The next version of Boabab will, to be able to move forward, introduce some breaking changes. As such, this version shall be the `1.0.0` one to respect semver principles.

Here is a list of things that will probably change so that we can discuss them before they are actually implemented.

**Note**: any development related to this `1.0.0` version will take place on a dedicated branch of the repository.

## Reference shifting

Reference shifting will become Baobab's default while the option should disappear altogether. This is necessary to ensure that easy "immutabley" comparisons are possible.

Related to #86.

## Dropping typology

Version `1.0.0` will remove its dependency to the `typology` library. Instead, the library will give access to a validation hook that anyone will be able to customize at will rather than being imposed a way to do so by Baobab. This should also make the library lighter (~4/5 ko AFAIK).

*Validate option proposition*

```js
var tree = new Baobab(data, {
  validate: function(oldState, newState, paths) {

    // Perform validation and return an error if invalid
    return new Error('invalid');
  }
})
```

It should be decided what a failed validation should trigger (an event, preventing a part of the transaction, the whole transaction etc.).

Related to #56.

## History at cursor level

History at tree level will disappear. Rather, for perf and legibility reasons, history methods will now belong to cursors. You'll still be able to record the whole's state history through the root cursor obviously.

```js
cursor.startRecording(maxRecords);
cursor.stopRecording();

cursor.hasHistory();
cursor.getHistory();

// Undo last step
cursor.undo();

// Undo n steps
cursor.undo(3);

// Same for redo
cursor.redo();
```

Related to #36 and #23.

## Dropping some options

* `clone`: will disappear because it is not very useful and makes the library's code more complex and fat for no good reason.
* `cloningFunction`: will disappear too for obvious reasons.
* `singletonCursors`: should disappear.
* `maxHistory`: will disappear for the reasons stated above.
* `typology`: useless if `typology` dependencies is removed.

Related to #98.

## React mixin

The react mixin will be simplified to reflect cursor's data directly on the components' state.

```jsx
var Component = React.createClass({
  mixins: [tree.mixin],

  // Single 'cursors' key, no more 'cursor' or 'cursors' etc.
  // Must be an object so values can be mapped to the component's state
  cursors: {

    // Either a path
    name: ['items', 0, 'name'],

    // Or a cursor
    surname: tree.select('items', 0, 'surname'),

    // Or a function so the cursor can be created dynamically with props, for instance
    points: function() {
      return this.props.givenCursor;
    }
  },

  // Accessing the cursors' values
  render: function() {
    return (
      <div>
        <p>{this.state.name} + ' ' + {this.state.surname}</p>
        <p onClick={this.handleClick}>{this.state.points} + ' points!'</p>
      </div>
    );
  },

  // Accessing the cursors
  handleClick: function() {

    // This is not necessarily a good idea to let your components handle the state updates by
    // themselves but you could do it if needed.
    this.cursors.name.edit('John');

    // You can also access the tree
    this.tree.get();
  }
});

// Should be possible to pass a single function to the `cursors` key and returning an object as a
// handy polymorphism.
var Component = React.createClass({
  mixins: [tree.mixin],
  cursors: function() {
    return {
      name: ['items', 0, 'name'],
      surname: tree.select('items', 0, 'surname'),
      points: this.propos.givenCursor
    };
  }
});
```

Alternatively, we should explore the ways to provide for a mixin that would not rely on a tree instance.

Related to #110 and #134

## Dynamic cursor reference paths

As proposed by @RJ in #128, dynamic cursor paths using themselves other cursor might be included in `1.0.0`.

```js
var tree = new Baobab({
  items: [
    'one',
    'two'
    'three'
  ],
  currentItem: 2
});

// Will resolve to the 'three' node
var cursor = tree.select('items', {$cursor: ['currentItem']});
```

## Facets proposal

Sometimes, you need to look at a specific view of your state and, what is more complex, listen to changes concerning this particular view. To achieve this particular point, I propose here to be able to create facets for your tree.

Facets would work exactly like React components except that they would listen the tree and return data rather than visual components. Also, you could listen to those facets and get their value exactly the same way as with the cursors.

*Example*

```js
var tree = new Baobab(

  // Data
  {
    projects: [
      {
        id: 1,
        name: 'Tezcatlipoca',
        user: 'John'
      },
      {
        id: 2,
        name: 'Huitzilopochtli',
        user: 'John'
      },
      {
        id: 3,
        name: 'Tlaloc',
        user: 'Jack'
      }
    ],
    currentProjectId: 1
  },

  // Options
  {
    facets: {

      // Filtered projects
      filteredProjects: {
        cursors: {
          projects: ['projects']
        },
        get: function(data) {
          return data.projects.filter(function(p) {
            return p.user === 'John';
          });
        }
      },

      // Current project's data
      currentProject: {
        cursors: {
          id: ['currentProjectId'],
          projects: ['projects']
        },
        get: function(data) {
          return _.first(data.projects, {id: data.id});
        }
      }
    }
  }
);

tree.facets.filteredProjects.get();
tree.facets.filteredProjects.on('update', fn);
```

This would be useful in cases you need to use this view in more than one part of your application. Else, doing the data filtering would rather belong to your components.

**Comment by RJ** [comment]
this all looks great 👍

I'll give the 1.0.0 branch a try asap

**Comment by Yomguithereal** [comment]
There is no such branch for the time being :). I haven't started coding it yet.

**Comment by christianalfoni** [comment]
Looking good :+1:  I just want to mention this issue, as this is an API changing version: https://github.com/Yomguithereal/baobab/issues/128#issuecomment-83851289

**Comment by jondlm** [comment]
This stuff looks great. We're starting to use Baobab at my work and we're loving it. These changes are totally in line with our collective reasoning about useful features. "Facets" in particular are something we've been thinking about. Having a "view" layer (akin to databases) on top of state is sweet.

**Comment by oskbor** [comment]
Awesome ideas, looking forward to trying them out

**Comment by ivan-kleshnin** [comment]
> These changes are totally in line with our collective reasoning about useful features. 
And with mine too, if that matters :)

**Comment by Yomguithereal** [comment]
@ivan-kleshnin, what do you think should be done in this regard? At the beginning, the `update` syntax was only meant to be private and used by the library internally for convenience but was exposed so that other could profit from it.

**Comment by ivan-kleshnin** [comment]
@Yomguithereal I worked with MongoDB alot and used to dislike it's query syntax with time. They constantly add new operators and still you always find yourself lacking one... I think Facebook proposal about that `update` was rather thoughtless because that... just doesn't fit with everything else. To insert you do LINQ, to update you pile those JSON queries :confused: to select and to remove you do LINQ again. Why? 

As an "anti-bonus" this `update` forced you to have strange `edit` for real `update`. Maybe naive but I would remove it entirely.

**Comment by ivan-kleshnin** [comment]
Also for now Baobab is not really datatype agnostic. Let's inspect different model representation formats. Many people tend to believe that models should be represented as Array while Object is actually more comfortable. Offsets are over-imperative.

```javascript
let Baobab = require("baobab");

let State= Baobab({
  robots: {
    listModels: [{id: "99"}],
    objModels: {"99": {id: "99"}},
  },
});

console.log("State/robots/listModels/0:", State.get("robots", "listModels", 0));
console.log("State/robots/objModels/99:", State.get("robots", "objModels", "99"));

console.log("State/robots/listModels/0:", State.get("robots", "listModels", {id: "99"}));
console.log("State/robots/objModels/{id: '99'}:", State.get("robots", "objModels", {id: "99"}));
```

```
State/robots/listModels/0: Object {id: "99"} // found
State/robots/objModels/99: Object {id: "99"} // found
State/robots/listModels/0: Object {id: "99"} // found
State/robots/objModels/{id: '99'}: undefined // not found
```

**Comment by Yomguithereal** [comment]
Development started under the [v1](https://github.com/Yomguithereal/baobab/tree/v1) branch.

**Comment by Yomguithereal** [comment]
The v1 is nearly complete concerning its core API. History, facets etc remain to be done but else than that, the library is usable. I'll soon rewrite the v1 documentation and start experimenting on react integration on a new `baobab-react` repository.

**Comment by christianalfoni** [comment]
Hi @Yomguithereal,

A couple of things I wanted to bring up:

1. I think it might be valuable to cache the facet "getter". So when the getter is triggered multiple times without any cursor change it does not do the calculation every time. Just a thought :-) Been playing around with it and really like the idea!

2. We have been playing around with a different idea at our offices. The idea is to be able to listen to "missing path" cursors. This is the scenario:

```js
var tree = new Baobab({
  users: {}
});

tree.select('users').on('missing', function (path) {
 
  // Create a temp value
  tree.select(path).edit({
   $isLoading: true
  });  
  
  // Do something to make the path valid, for example
  // requesting data from the backend
  var request = ajax.get(path);

  request.done(function (response) {
    tree.select(path).edit(response);  
  }).catch(function (error) {
    tree.select(path).edit({
      $error: error
    });
  });
});

var user123 = tree.select('users', '123');
user123.get(); // {$isLoading: true}
// later
user123.get(); // {foo: 'bar'}
```

The reason this came up is because it is very typical that data from the server references IDs which you need to collect. By registering these "missing" listeners you are able to evaluate values that the UI wants and can make sure that they do get a value, and an eventual value if you want.

Does this makes sense? :-)

**Comment by Yomguithereal** [comment]
Hi @christianalfoni,

1. This is indeed what I had in mind ;-).

2. We were asking this same question with @jacomyal, lately. The thing which might get tricky here is to be able to define what means a path is "missing". `undefined` value? Non-existent path etc. But this would clearly be a nice thing to cover indeed. Moreover, sometimes you want this missing event to be "lazy", so your UI loads the data only if this is actually needed.

**Comment by christianalfoni** [comment]
@Yomguithereal 

1. Awesome! :-)

2. Yeah, that is a very good point :-) In my opinion I would suggest it to be a missing "key". So not `path !== undefined`, but `lastPathValue in leadingPathValue`, converts to, `'123' in 'users'`. So an undefined value would still be a valid value, it is the existence of the key I think you are after.

Hm, did not quite understand the "lazy" thingy. You got time for a small example? :-)

**Comment by Yomguithereal** [comment]
The missing key seems to be rather rational but I often initialize my tree with a full object containing every "possible" key and setting the ones that will change later to something like `undefined` or `null`. I do that to be able to visualize easily what could be my tree fully "loaded". In this case, the missing event would not work and this is sad.

Concerning the lazy part: imagine a UI with several routes. Those routes might need different data contained in different places of the tree. I don't want to load the data I don't need for the current displayed route. The missing event would have to be triggered on the first get at the given path and returning an empty value for instance, to be lazy. But maybe this is what you initially meant.

**Comment by christianalfoni** [comment]
Ah, yeah, that is a very good point! Hm hm, have to let this one sink in ;-)

Aha, I see. Actually I was considering that the event would trigger when registering the cursor. So you could have as much undefined and get you want, but it is when you are interested in "changes" to the value you get that functionality.

Typically a React component:
```js
var MyComponent = React.createClass({
  mixin: [state.mixin],
  cursors: function () {
    user: ['users', this.props.userId]
  },
  render: function () {
    return <h1>{this.state.user.name}</h1>
  }
});
```
As soon as the cursor is registered you kinda say: "Okay, I am definitely interested in this value, but give it when you are good and ready, optionally give me a temp value back immediately".. hope I am understanding you correctly with the lazyness :-)

**Comment by christianalfoni** [comment]
You know, I started thinking.... I often do ;-)

What about just listening for new cursors added? That way you could yourself evaluate what cursors to act upon:

```js
var tree = new Baobab({
  users: {}
});

tree.select('users').on('cursor', function (path) {
 
  if (tree.path.length === 2 && tree.get(path) === undefined) {
    // Do some logic to resolve a value
  }

});
```
This way it is more clear that the event will trigger when a cursor is registered. It also gives a lot more flexibility as it lets you evaluate the path, evaluate what value to act upon, or lack there of.

Hm hm, what do you think @Yomguithereal?

**Comment by bebraw** [comment]
It could be useful if you had `cursor.clearHistory();` in addition to the planned methods. There may be cases when you may want to invalidate whole undo stack and it doesn't seem like a tough one to add.

**Comment by Yomguithereal** [comment]
Indeed. You can currently achieve the same effect by doing

```js
cursor.stopRecording();
cursor.startRecording();
```

but `cursor.clearHistory()` is clearly more straightforward and enable one to stopRecording while keeping the records.

**Comment by Yomguithereal** [comment]
Proposition of react integration [here](https://github.com/Yomguithereal/baobab-react).

**Comment by Yomguithereal** [comment]
I've just released Baobab `v1.0.0-rc1` on npm.

You can now install v1 through:

```
npm install baobab@1.0.0-rc1
# or
npm install baobab@beta
```

Documentation is available on the related branch [here](https://github.com/Yomguithereal/baobab/tree/v1#baobab).

Is there any brave early adopters willing to switch to v1 so we can advance further :) ?

**Comment by Yomguithereal** [comment]
Version `1.0.0` is now released. I will therefore, ere you noble assembly, close this issue.
